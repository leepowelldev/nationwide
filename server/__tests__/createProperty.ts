import { gql } from 'apollo-server';
import { mockEmptyCollection } from './__utils__/mockCollections.js';
import setup from './__utils__/setup.js';

jest.mock('uuid', () => ({
  v4: jest.fn(() => 'XXXXX-XXXXX-XXXXX'),
}));

const query = gql`
  mutation ($input: PropertyInput!) {
    createProperty(input: $input) {
      id
      address
      type
      bedrooms
    }
  }
`;

describe('Mutation: createProperty', () => {
  test('should create property', async () => {
    const { server, collection } = setup(mockEmptyCollection);
    const input = {
      address: 'A house somewhere',
      type: 'HOUSE',
      bedrooms: 3,
    };

    // Check collection has default size
    expect(collection.size).toStrictEqual(0);

    const res = await server.executeOperation({
      query,
      variables: {
        input,
      },
    });

    // Check collection has changed size
    expect(collection.size).toStrictEqual(1);
    expect(res).toMatchSnapshot();
  });

  test('should error when property address is too short (yup schema)', async () => {
    const { server, collection } = setup(mockEmptyCollection);

    // Check collection has default size
    expect(collection.size).toStrictEqual(0);

    const res = await server.executeOperation({
      query,
      variables: {
        input: {
          address: 'A hou',
          type: 'HOUSE',
          bedrooms: 3,
        },
      },
    });

    // Check collection has not changed size
    expect(collection.size).toStrictEqual(0);
    expect(JSON.stringify(res.errors?.[0])).toMatchSnapshot();
    expect(res).toMatchSnapshot();
  });

  test('should error when property input is missing field (internal)', async () => {
    const { server, collection } = setup(mockEmptyCollection);

    expect(collection.size).toStrictEqual(0);

    const res = await server.executeOperation({
      query,
      variables: {
        input: {
          address: 'A house somewhere',
          bedrooms: 3,
        },
      },
    });

    expect(collection.size).toStrictEqual(0);
    expect(JSON.stringify(res.errors?.[0])).toMatchSnapshot();
    expect(res).toMatchSnapshot();
  });
});
