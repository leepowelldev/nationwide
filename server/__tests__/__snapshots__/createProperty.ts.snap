// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation: createProperty should create property 1`] = `
Object {
  "data": Object {
    "createProperty": Object {
      "address": "A house somewhere",
      "bedrooms": 3,
      "id": "XXXXX-XXXXX-XXXXX",
      "type": "HOUSE",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createProperty should error when property address is too short (yup schema) 1`] = `"{\\"message\\":\\"address must be at least 10 characters\\",\\"locations\\":[{\\"line\\":2,\\"column\\":3}],\\"path\\":[\\"createProperty\\"],\\"extensions\\":{\\"code\\":\\"BAD_USER_INPUT\\"}}"`;

exports[`Mutation: createProperty should error when property address is too short (yup schema) 2`] = `
Object {
  "data": Object {
    "createProperty": null,
  },
  "errors": Array [
    [GraphQLError: address must be at least 10 characters],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createProperty should error when property input is missing field (internal) 1`] = `"{\\"message\\":\\"Variable \\\\\\"$input\\\\\\" got invalid value { address: \\\\\\"A house somewhere\\\\\\", bedrooms: 3 }; Field \\\\\\"type\\\\\\" of required type \\\\\\"PropertyType!\\\\\\" was not provided.\\",\\"locations\\":[{\\"line\\":1,\\"column\\":11}],\\"extensions\\":{\\"code\\":\\"BAD_USER_INPUT\\"}}"`;

exports[`Mutation: createProperty should error when property input is missing field (internal) 2`] = `
Object {
  "data": undefined,
  "errors": Array [
    [UserInputError: Variable "$input" got invalid value { address: "A house somewhere", bedrooms: 3 }; Field "type" of required type "PropertyType!" was not provided.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
